docs/08-gitcoin-identity-staking-management.md

```diff
@@ -0,0 +1,105 @@
+# Rolling 3-24-Months Identity Staking and re-staking
+
+This describes the contracts and state used for the rolling 3-24 months identity staking and re-staking.
+
+## Data Structures
+
+```solidity
+// uint256 stakeCount;
+uint256 stakeCount;
+
+// mapping(selfStakerAddress => uint[]) selfStakes;
+mapping(address => uint256[]) selfStakes;
+
+// mapping(stakerAddress => mapping(stakeeAddress => uint[])) communityStakes;
+mapping(address => mapping(address => uint256[])) communityStakes;
+
+mapping(uint256 => Stake) stakes;
+
+// mapping(stakerAddress => stakeeAddress[]) addressesStakedOnByUser;
+mapping(address => address[]) addressesStakedOnByUser;
+
+// mapping(stakeeAddress => stakerAddress[]) addressesStakingOnUser;
+mapping(address => address[]) addressesStakingOnUser;
+
+struct Stake {
+  uint256 amount;
+  uint256 unlockTime;
+}
+```
+
+We considered using an enumerable set instead of lists to track Stake IDs. Although all operations besides writing were more efficient with an enumerable set, the difference when writing a new value/staking was considerably different, and assuming that staking would be the most common operation, we decided to use lists instead. Here is a comparison of the gas costs for each operation: https://miro.com/app/board/uXjVNFnedyI=/
+
+## Methods
+
+Relevant methods and their responsibilities
+
+```solidity
+function selfStake(address staker, Stake stake) external
```

Comment:
What about defining this function like this:
function `selfStake(uint256 amount, uint64 duration)`

Then you can create the instance of the `Stake` structure like this:
- set amount to what user has specified
- calculate `lockTime` and `unlockTime`: `lockTime` will be `block.timestamp`, and you can calculate `unlockTime` by adding the duration to that

And you don't need the address since it should always be `msg.sender`, right?

Reply:
Good call with msg.sender, that has been updated. My thought with passing the unlockTime instead calculating unlock and lock, is that you can calculate that offchain and save a bit of gas. Either way we will have to verify that the unlock or duration is within an appropriate range